### PR TITLE
[cker] Introduce Subtract gradient kernel

### PR DESCRIPTION
**Pull request by Aeren1564**
This commit introduces Subtract gradient kernel.

ONE-DCO-1.0-Signed-off-by: YongHyun An <yonghyunz.an@samsung.com>

---
From draft #12417

### CODE REVIEW

**Comment by Aeren1564** [comment]
This implementation has been verified on https://github.com/Samsung/ONE/pull/12417#issuecomment-1911435577.

**Review by jyoungyun — APPROVED** [review]
LGTM

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h [hunk 1]
```diff
@@ -51,6 +51,15 @@ void BinaryArithmeticGrad(const Shape &incoming_shape, const T *incoming_data,
     break;
 
     case ArithmeticType::kSub:
+    {
+      BroadcastTo(incoming_shape, const_cast<T *>(incoming_data), lhs_grad_shape, lhs_grad_data);
+
+      auto const in_map = MapAsMatrixWithLastDimAsRows(incoming_data, incoming_shape);
+      auto rhs_grad_map = MapAsMatrixWithLastDimAsRows(rhs_grad_data, rhs_grad_shape);
+      rhs_grad_map = -in_map;
```
Comment (zetwhite) [inline compute/cker/include/cker/train/operation/BinaryArithmetic.h:59]: Thinking about `sub` gradient, I thought that for RHS multiplying `-1` is enough.

But here, Functions like `MapAsMatrixWithLastDimAsRows` is called.
If you don't mind, could you explain a bit - why this function is called? (I'm not familiar with Eigen :sweat_smile: )

Comment (Aeren1564) [reply compute/cker/include/cker/train/operation/BinaryArithmetic.h:59]: > Thinking about `sub` gradient, I thought that for RHS multiplying `-1` is enough.
> 
> But here, Functions like `MapAsMatrixWithLastDimAsRows` is called. If you don't mind, could you explain a bit - why this function is called? (I'm not familiar with Eigen 😅 )

To my understanding, `MapAsMatrixWithLastDimAsRows` is generates a multidimensional view of underlying data, and calling it itself does not modify anything. I thought that Implementing this way makes it clear that we're doing elementwise negation on the multidimensional data.

File: compute/cker/src/train/BinaryArithmetic.test.cc [hunk 2]
```diff
@@ -58,3 +58,43 @@ TEST(CKer_Operation, AddGrad)
...
+  {
+    // Shape: {2, 3}
+    std::vector<float> incoming_backward = {-2, 3, -4, 5, -6, 7};
+    std::vector<float> expected_lhs_backward = incoming_backward;
+    std::vector<float> expected_rhs_backward = incoming_backward;
+    std::vector<float> lhs_backward(6);
+    std::vector<float> rhs_backward(6);
+
+    nnfw::cker::train::BinaryArithmeticGrad(nnfw::cker::Shape{2, 3}, incoming_backward.data(),
+                                            nnfw::cker::Shape{2, 3}, lhs_backward.data(),
```
Comment (glistening) [inline compute/cker/src/train/BinaryArithmetic.test.cc:73]: @Aeren1564 Did you test with different shape of `lhs` and `rhs`. For example, it would be good to have a test case with `lhs.shape = (1,3)`. Also, what happens when it is not correct shape (that is `lhs.shape = (3,3)`) ? Do we assume its shape is always correct because forward pass will be done before backward pass?

Comment (Aeren1564) [reply compute/cker/src/train/BinaryArithmetic.test.cc:73]: Is it possible to have `lhs` and `rhs` with different shapes?

I assumed that we'll always have valid shape because [tf.keras.layers.Subtract](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Subtract) specification stated that

```It takes as input a list of tensors of size 2, both of the same shape, and returns a single tensor, (inputs[0] - inputs[1]), also of the same shape.```

so I thought that we can't have input with different shapes.

Comment (ragmani) [reply compute/cker/src/train/BinaryArithmetic.test.cc:73]: > Is it possible to have lhs and rhs with different shapes?

Yes, it is possible. But, considering the broadcasting of Eigen(https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html), The current kernel does not support broadcasting yet.

Comment (ragmani) [reply compute/cker/src/train/BinaryArithmetic.test.cc:73]: By the way, I guess the current kernel supports only 2D. If not, could you add tests with rank higher than 3?

Comment (Aeren1564) [reply compute/cker/src/train/BinaryArithmetic.test.cc:73]: 1. The following check has been added.
```cpp
  if (lhs_grad_shape != rhs_grad_shape)
    throw std::runtime_error{"Shape of LHS and RHS must match"};
```

2. Unittest for distinct shape has been added.

> By the way, I guess the current kernel supports only 2D. If not, could you add tests with rank higher than 2?

3. Unittest for dimension $2x2x2$ has been added.

Comment (ragmani) [reply compute/cker/src/train/BinaryArithmetic.test.cc:73]: Could you also add checking shape of incoming?

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h [hunk 3]
```diff
@@ -51,6 +54,15 @@ void BinaryArithmeticGrad(const Shape &incoming_shape, const T *incoming_data,
...
```
Comment (ragmani) [inline compute/cker/include/cker/train/operation/BinaryArithmetic.h:63]: (Optional)
If you are not going to support broadcast in this kernel, this code may be replaced with as below.
```suggestion
      auto const in_map = MapAsVector(incoming_data, incoming_shape);
      auto rhs_grad_map = MapAsVector(rhs_grad_data, rhs_grad_shape);
      rhs_grad_map.array() = -in_map.array();
```
I'm not sure, but vectors may be slightly faster than matrixes.

File: compute/cker/src/train/BinaryArithmetic.test.cc (hunk 2 above)
Comment (Aeren1564) [reply compute/cker/src/train/BinaryArithmetic.test.cc:73]: The check has been modified to

```cpp
  if (lhs_grad_shape != rhs_grad_shape || lhs_grad_shape != incoming_shape ||
      rhs_grad_shape != incoming_shape)
    throw std::runtime_error{"Shape of LHS, RHS, and incoming must match"};
```

**Review by ragmani — APPROVED** [review]
LGTM

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h (hunk 1 above)
Comment (zetwhite) [reply compute/cker/include/cker/train/operation/BinaryArithmetic.h:63]: Oh I see.  

(suggestion) 

I read the code and It looks like `MapAsMatrixWithLastDimAsRows` creates a data view for easy access alongside the last dimensions. If I understand correctly, it looks like there is no access alongside specific dimensions. 

If so, How about using `MapAsVector` instead of `MapAsMatrixWithDimAsRows`? 
Maybe like this..? 
```cpp 
      const auto incoming_map =  MapAsVector(incoming_data, incoming_shape);
      auto rhs_grad_map = MapAsVector(rhs_grad_data, rhs_grad_shape);
      rhs_grad_map = -1 * incoming_map.array() ; 
```

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h (hunk 3 above)
Comment (Aeren1564) [reply compute/cker/include/cker/train/operation/BinaryArithmetic.h]: Updated

Comment (zetwhite) [reply compute/cker/include/cker/train/operation/BinaryArithmetic.h]: (not for this PR) 

https://github.com/Samsung/ONE/blob/8185cb17da4f1e6d84bb14758fec693dd9c38c62/compute/cker/include/cker/eigen/Utils.h#L37-L47

Hmm.. It looks like `MapAsVector` just creates another matrix .. `Eigen::Matrix<Scaler, N, 1>`.
We might need to introduce Util which directly creates `Eigen::Aray()` instead of creating `Matrix`.

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h (hunk 1 above)
Comment (Aeren1564) [reply compute/cker/include/cker/train/operation/BinaryArithmetic.h]: Updated.

**Review by zetwhite — APPROVED** [review]
LGTM :+1:

**Review by ragmani — APPROVED** [review]

File: compute/cker/include/cker/train/operation/BinaryArithmetic.h [hunk 4]
```diff
@@ -51,6 +55,15 @@ void BinaryArithmeticGrad(const Shape &incoming_shape, const T *incoming_data,
     break;
 
     case ArithmeticType::kSub:
+    {
+      BroadcastTo(incoming_shape, const_cast<T *>(incoming_data), lhs_grad_shape, lhs_grad_data);
```
[the remark this answers is not in the document]
Comment (ragmani) [reply compute/cker/include/cker/train/operation/BinaryArithmetic.h:59]: (A) I guess this code is necessary to back-propagate incoming data to lhs gradient data.

**Review by jyoungyun — APPROVED** [review]
LGTM

**Review by glistening — APPROVED** [review]
LGTM
